hummingbot/wallet/ethereum/watcher/websocket_watcher.py

```diff
@@ -0,0 +1,169 @@
+
+    _nbw_logger: Optional[HummingbotLogger] = None
+
+    @classmethod
+    def logger(cls) -> HummingbotLogger:
+        if cls._nbw_logger is None:
+            cls._nbw_logger = logging.getLogger(__name__)
+        return cls._nbw_logger
+
+    @property
+    def block_number(self) -> int:
+        return self._current_block_number
+
+    @property
+    def block_cache(self) -> Dict[HexBytes, AttributeDict]:
+        cache_dict: Dict[HexBytes, AttributeDict] = dict([(key, self._block_cache[key])
+                                                          for key in self._block_cache.keys()])
+        return cache_dict
+
+    async def start_network(self):
+        if self._fetch_new_blocks_task is not None:
+            await self.stop_network()
+        else:
+            try:
+                self._current_block_number = await self.call_async(getattr, self._w3.eth, "blockNumber")
+            except asyncio.CancelledError:
+                raise
+            except Exception:
+                self.logger().network("Error fetching newest Ethereum block number.",
+                                      app_warning_msg="Error fetching newest Ethereum block number. "
+                                                      "Check Ethereum node connection",
+                                      exc_info=True)
+            await self.connect()
+            await self.subscribe(["newHeads"])
+            self._fetch_new_blocks_task: asyncio.Task = safe_ensure_future(self.fetch_new_blocks_loop())
+            self._network_on = True
+
+    async def stop_network(self):
+        if self._fetch_new_blocks_task is not None:
+            await self.disconnect()
+            self._fetch_new_blocks_task.cancel()
+            self._fetch_new_blocks_task = None
+            self._network_on = False
+
+    async def connect(self):
+        try:
+            self._client = await websockets.connect(uri=self._websocket_url)
+            return self._client
+        except Exception as e:
+            self.logger().network(f"ERROR in connection: {e}")
+
+    async def disconnect(self):
+        try:
+            await self._client.close()
+        except Exception as e:
+            self.logger().network(f"ERROR in disconnection: {e}")
+
+    async def _send(self, emit_data) -> int:
+        self._nonce += 1
+        emit_data["id"] = self._nonce
+        await self._client.send(ujson.dumps(emit_data))
+        return self._nonce
+
+    async def subscribe(self, params) -> bool:
+        emit_data = {
+            "method": "eth_subscribe",
+            "params": params
+        }
+        nonce = await self._send(emit_data)
+        raw_message = await self._client.recv()
+        if raw_message is not None:
+            resp = ujson.loads(raw_message)
+            if resp.get("id", None) == nonce:
+                self._node_address = resp.get("result")
+                return True
+        return False
+
+    async def unsubscribe(self, params) -> bool:
```

Comment:
Do we need this? Is it used anywhere?

Reply:
It's not used, but would be necessary if this became generalized for handling WebSockets. I can remove it